¿Cómo interpretar los negativos en la columna DistanceFromHome? Puede haber empleados de on site en otras empresas haciendo labor de consultoría.
¿La Distancia está en km?
¿Qué es Monthlyrate?

1. Introducción

📌 Análisis Exploratorio de Datos para ABC Corporation

📝 Descripción: El objetivo de este análisis es comprender mejor la estructura y el contenido de los datos de empleados de ABC Corporation, identificar problemas de calidad de datos y realizar las transformaciones necesarias para preparar los datos para un análisis más profundo.

In [1]:
#2. Carga de Librerías y Datos 📚
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("HR RAW DATA.csv", index_col=0)
df.tail()

Age Attrition BusinessTravel DailyRate Department  DistanceFromHome  \
1609  36       Yes  travel_rarely    530,0$        NaN                 3   
1610  45        No     non-travel    805,0$        NaN                 4   
1611  39        No  travel_rarely    903,0$        NaN               -13   
1612  36        No     non-travel   1229,0$        NaN                 8   
1613  46        No            NaN    566,0$        NaN                 7   

      Education    EducationField  employeecount employeenumber  ...  \
1609          1     Life Sciences              1          967,0  ...   
1610          2               NaN              1          972,0  ...   
1611          5               NaN              1            NaN  ...   
1612          4  Technical Degree              1          990,0  ...   
1613          2           Medical              1            NaN  ...   

      YearsAtCompany  YearsInCurrentRole YearsSinceLastPromotion  \
1609              16                 NaN                       3   
1610               9                 NaN                       0   
1611               8                 NaN                       0   
1612               7                 NaN                       0   
1613               8                 NaN                       0   

      YEARSWITHCURRMANAGER  SameAsMonthlyIncome DateBirth       Salary  \
1609                     7              10325,0      1987  1000000000$   
1610                     8               4447,0      1978  1000000000$   
1611                     7                  NaN      1984  1000000000$   
1612                     7                  NaN      1987  1000000000$   
1613                     7              10845,0      1977  1000000000$   

     RoleDepartament NUMBERCHILDREN  RemoteWork  
1609             NaN            NaN           0  
1610             NaN            NaN           1  
1611             NaN            NaN         Yes  
1612             NaN            NaN        True  
1613             NaN            NaN           0  

[5 rows x 41 columns]

In [3]:
#3. Análisis Inicial del Conjunto de Datos 🔍
#📊 Resumen del DataFrame:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1614 entries, 0 to 1613
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1614 non-null   object 
 1   Attrition                 1614 non-null   object 
 2   BusinessTravel            842 non-null    object 
 3   DailyRate                 1614 non-null   object 
 4   Department                302 non-null    object 
 5   DistanceFromHome          1614 non-null   int64  
 6   Education                 1614 non-null   int64  
 7   EducationField            869 non-null    object 
 8   employeecount             1614 non-null   int64  
 9   employeenumber            1183 non-null   object 
 10  EnvironmentSatisfaction   1614 non-null   int64  
 11  Gender                    1614 non-null   int64  
 12  HourlyRate                1614 non-null   object 
 13  JobInvolvement            1614 non-null   int64  
 14  JobLevel     

In [4]:
#📈 Descripción Estadística:
df.describe(include="O").T

count unique                       top  freq
Age                  1614     54                        35    84
Attrition            1614      2                        No  1355
BusinessTravel        842      3             travel_rarely   586
DailyRate            1614    849                      nan$   124
Department            302      3   Research & Development    196
EducationField        869      6             Life Sciences   349
employeenumber       1183   1079                     482,0     2
HourlyRate           1614     72             Not Available    84
JobRole              1614   1579                  mANager      4
MaritalStatus         963      5                   Married   404
MonthlyIncome         771    668                    6347,0     4
Over18                713      1                         Y   713
OverTime              938      2                        No   682
PerformanceRating    1419      2                       3,0  1205
StandardHours         419      1                      80,0   419
TOTALWORKINGYEARS    1088     40                      10,0   144
WORKLIFEBALANCE      1506      4                       3,0   913
YearsInCurrentRole     34     10                       2,0    11
SameAsMonthlyIncome   771    668                    6347,0     4
Salary               1614      1               1000000000$  1614
RoleDepartament       302    301        MaNAgeR  -  Sales      2
RemoteWork           1614      5                         1   360

In [5]:
df[df["DistanceFromHome"]<0]["DistanceFromHome"].min()

-49

In [6]:
df[df["DistanceFromHome"]>0]["DistanceFromHome"].max()

29

In [7]:
len(df["YearsInCurrentRole"].dropna())

34

In [8]:
pd.set_option('display.max_columns', None)

In [9]:
df["SameAsMonthlyIncome"].describe()

count        771
unique       668
top       6347,0
freq           4
Name: SameAsMonthlyIncome, dtype: object

In [10]:
df["MonthlyIncome"].describe()

count        771
unique       668
top       6347,0
freq           4
Name: MonthlyIncome, dtype: object

In [11]:
df.loc[:,["DailyRate", "HourlyRate", "MonthlyIncome", "MonthlyRate", "PercentSalaryHike", "StandardHours"]]

DailyRate HourlyRate MonthlyIncome  MonthlyRate  PercentSalaryHike  \
0       684,0$         51       19537,0         6462                 13   
1       699,0$         65       19999,0         5678                 14   
2       532,0$         58       19232,0         4933                 11   
3       359,0$         82       17169,0        26703                 19   
4      1319,0$         45           NaN         7739                 12   
...        ...        ...           ...          ...                ...   
1609    530,0$         51       10325,0         5518                 11   
1610    805,0$         57        4447,0        23163                 12   
1611    903,0$         41           NaN         2560                 18   
1612   1229,0$         84           NaN        25952                 13   
1613    566,0$         75       10845,0        24208                 13   

     StandardHours  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
...            ...  
1609           NaN  
1610           NaN  
1611           NaN  
1612           NaN  
1613           NaN  

[1614 rows x 6 columns]

In [12]:
df["Over18"].isnull().sum()

901

In [13]:
df["Over18"].value_counts()

Over18
Y    713
Name: count, dtype: int64

In [14]:
df.drop(labels="SameAsMonthlyIncome", axis=1, inplace=True)

In [15]:
df["DistanceFromHome"].isna().sum()

0

In [16]:
df["DistanceFromHome"].value_counts()

DistanceFromHome
 2     217
 1     203
 10     86
 9      85
 8      81
      ... 
-28      2
-34      2
-43      2
-39      1
-40      1
Name: count, Length: 69, dtype: int64

In [17]:
df.iloc[:,5:]

DistanceFromHome  Education    EducationField  employeecount  \
0                    6          3               NaN              1   
1                    1          4     Life Sciences              1   
2                    4          2  Technical Degree              1   
3                    2          4           Medical              1   
4                    3          3  Technical Degree              1   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences              1   
1610                 4          2               NaN              1   
1611               -13          5               NaN              1   
1612                 8          4  Technical Degree              1   
1613                 7          2           Medical              1   

     employeenumber  EnvironmentSatisfaction  Gender HourlyRate  \
0             162,0                        1       0         51   
1             259,0                        3       0         65   
2             319,0                        3       0         58   
3               NaN                        1       1         82   
4               NaN                        1       1         45   
...             ...                      ...     ...        ...   
1609          967,0                        3       0         51   
1610          972,0                        3       0         57   
1611            NaN                       13       0         41   
1612          990,0                        1       0         84   
1613            NaN                        4       0         75   

      JobInvolvement  JobLevel                   JobRole  JobSatisfaction  \
0                  3         5        resEArch DIREcToR                 3   
1                  2         5                  ManAGeR                 3   
2                  3         5                  ManaGER                 4   
3                  3         4        ReseArCH DIrECtOr                 3   
4                  4         4          sAleS EXECUtIve                 1   
...              ...       ...                       ...              ...   
1609               2         3          saLEs ExeCUTiVe                 4   
1610               3         2    LAboRaTOry tECHNiCIAn                 2   
1611               4         3          sAlES ExECUTivE                 3   
1612               3         2          SaLes ExecUtIVe                 4   
1613               3         3   mAnUfactURInG DiRECTOr                 3   

     MaritalStatus MonthlyIncome  MonthlyRate  NUMCOMPANIESWORKED Over18  \
0              NaN       19537,0         6462                   7      Y   
1              NaN       19999,0         5678                   0    NaN   
2          Married       19232,0         4933                   1    NaN   
3          Married       17169,0        26703                   3      Y   
4         Divorced           NaN         7739                   2      Y   
...            ...           ...          ...                 ...    ...   
1609       Married       10325,0         5518                   1      Y   
1610           NaN        4447,0        23163                   1    NaN   
1611        Single           NaN         2560                   0    NaN   
1612      Divorced           NaN        25952                   4    NaN   
1613           NaN       10845,0        24208                   6      Y   

     OverTime  PercentSalaryHike PerformanceRating  RelationshipSatisfaction  \
0          No                 13               3,0                         3   
1         NaN                 14               3,0                         1   
2          No                 11               3,0                         4   
3         NaN                 19               3,0                         2   
4          No                 12               3,0                         4   
...       ...                ...          

In [18]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'employeecount',
       'employeenumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NUMCOMPANIESWORKED',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TOTALWORKINGYEARS', 'TrainingTimesLastYear', 'WORKLIFEBALANCE',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YEARSWITHCURRMANAGER', 'DateBirth', 'Salary', 'RoleDepartament',
       'NUMBERCHILDREN', 'RemoteWork'],
      dtype='object')

In [19]:
eliminar = ['employeecount','Over18','Salary']
df_limpio = df.drop(columns=eliminar)

In [20]:
df_limpio

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField employeenumber  \
0                    6          3               NaN          162,0   
1                    1          4     Life Sciences          259,0   
2                    4          2  Technical Degree          319,0   
3                    2          4           Medical            NaN   
4                    3          3  Technical Degree            NaN   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences          967,0   
1610                 4          2               NaN          972,0   
1611               -13          5               NaN            NaN   
1612                 8          4  Technical Degree          990,0   
1613                 7          2           Medical            NaN   

      EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                           1       0         51               3         5   
1                           3       0         65               2         5   
2                           3       0         58               3         5   
3                           1       1         82               3         4   
4                           1       1         45               4         4   
...                       ...     ...        ...             ...       ...   
1609                        3       0         51               2         3   
1610                        3       0         57               3         2   
1611                       13       0         41               4         3   
1612                        1       0         84               3         2   
1613                        4       0         75               3         3   

                       JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0           resEArch DIREcToR                 3           NaN       19537,0   
1                     ManAGeR                 3           NaN       19999,0   
2                     ManaGER                 4       Married       19232,0   
3           ReseArCH DIrECtOr                 3       Married       17169,0   
4             sAleS EXECUtIve                 1      Divorced           NaN   
...                        ...              ...           ...           ...   
1609          saLEs ExeCUTiVe                 4       Married       10325,0   
1610    LAboRaTOry tECHNiCIAn                 2           NaN        4447,0   
1611          sAlES ExECUTivE                 3        Single           NaN   
1612          SaLes ExecUtIVe                 4      Divorced           NaN   
1613   mAnUfactURInG DiRECTOr                 3           NaN       10845,0   

      MonthlyRate  NUMCOMPANIESWORKED OverTime  PercentSalaryHike  \
0            6462                   7       No                 13   
1            5678                   0      NaN                 14   
2            4933                   1       No                 11   
3           26703                   3      NaN                 19   
4            7739                   2       No          

In [21]:
cambiar_nombre = {"employeenumber": "EmployerNumber", "NUMCOMPANIESWORKED": "NumCompaniesWorked", "TOTALWORKINGYEARS": "TotalWorkingYeras", "WORKLIFEBALANCE": "WorkLifeBalance",
                  "YEARSWITHCURRMANAGER": "YearsWithCurrManager", "NUMBERCHILDREN": "NumberChildren"}

df_limpio.rename(columns=cambiar_nombre, inplace=True)
df_limpio

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField EmployerNumber  \
0                    6          3               NaN          162,0   
1                    1          4     Life Sciences          259,0   
2                    4          2  Technical Degree          319,0   
3                    2          4           Medical            NaN   
4                    3          3  Technical Degree            NaN   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences          967,0   
1610                 4          2               NaN          972,0   
1611               -13          5               NaN            NaN   
1612                 8          4  Technical Degree          990,0   
1613                 7          2           Medical            NaN   

      EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                           1       0         51               3         5   
1                           3       0         65               2         5   
2                           3       0         58               3         5   
3                           1       1         82               3         4   
4                           1       1         45               4         4   
...                       ...     ...        ...             ...       ...   
1609                        3       0         51               2         3   
1610                        3       0         57               3         2   
1611                       13       0         41               4         3   
1612                        1       0         84               3         2   
1613                        4       0         75               3         3   

                       JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0           resEArch DIREcToR                 3           NaN       19537,0   
1                     ManAGeR                 3           NaN       19999,0   
2                     ManaGER                 4       Married       19232,0   
3           ReseArCH DIrECtOr                 3       Married       17169,0   
4             sAleS EXECUtIve                 1      Divorced           NaN   
...                        ...              ...           ...           ...   
1609          saLEs ExeCUTiVe                 4       Married       10325,0   
1610    LAboRaTOry tECHNiCIAn                 2           NaN        4447,0   
1611          sAlES ExECUTivE                 3        Single           NaN   
1612          SaLes ExecUtIVe                 4      Divorced           NaN   
1613   mAnUfactURInG DiRECTOr                 3           NaN       10845,0   

      MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike  \
0            6462                   7       No                 13   
1            5678                   0      NaN                 14   
2            4933                   1       No                 11   
3           26703                   3      NaN                 19   
4            7739                   2       No          

In [22]:
df_limpio["JobRole"] = df_limpio["JobRole"].str.title()
df_limpio.head()


Age Attrition BusinessTravel DailyRate                Department  \
0  51        No            NaN    684,0$                       NaN   
1  52        No            NaN    699,0$                       NaN   
2  42        No  travel_rarely    532,0$   Research & Development    
3  47        No  travel_rarely    359,0$                       NaN   
4  46        No            NaN   1319,0$                       NaN   

   DistanceFromHome  Education    EducationField EmployerNumber  \
0                 6          3               NaN          162,0   
1                 1          4     Life Sciences          259,0   
2                 4          2  Technical Degree          319,0   
3                 2          4           Medical            NaN   
4                 3          3  Technical Degree            NaN   

   EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                        1       0         51               3         5   
1                        3       0         65               2         5   
2                        3       0         58               3         5   
3                        1       1         82               3         4   
4                        1       1         45               4         4   

               JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0   Research Director                 3           NaN       19537,0   
1             Manager                 3           NaN       19999,0   
2             Manager                 4       Married       19232,0   
3   Research Director                 3       Married       17169,0   
4     Sales Executive                 1      Divorced           NaN   

   MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike  \
0         6462                   7       No                 13   
1         5678                   0      NaN                 14   
2         4933                   1       No                 11   
3        26703                   3      NaN                 19   
4         7739                   2       No                 12   

  PerformanceRating  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0               3,0                         3           NaN                 0   
1               3,0                         1           NaN                 1   
2               3,0                         4           NaN                 0   
3               3,0                         2           NaN                 2   
4               3,0                         4           NaN                 1   

  TotalWorkingYeras  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0               NaN                      5             3,0              20   
1              34,0                      5             3,0              33   
2              22,0                      3             NaN              22   
3               NaN                      2             NaN              20   
4               NaN                      5             3,0              19   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
0                NaN                       15                    15   
1                NaN                       11                     9   
2                NaN                       11                    15   
3                NaN                        5                     6   
4                NaN                        2                     8   

   DateBirth                       RoleDepartament  NumberChildren RemoteWork  
0       1972                                   NaN             NaN        Yes  
1       1971                                   NaN             NaN          1  
2       1981   ManaGER  -  Research & Development              NaN          1  
3       1976                                   NaN             NaN      False  
4       1977                                   NaN             NaN          0

In [23]:
df_limpio["MaritalStatus"].value_counts()

MaritalStatus
Married     404
Single      325
Divorced    188
Marreid      35
divorced     11
Name: count, dtype: int64

In [24]:
df_limpio["MaritalStatus"] = df_limpio["MaritalStatus"].str.title()

df_limpio["MaritalStatus"] = df_limpio["MaritalStatus"].str.replace("Marreid", "Married")

In [25]:
df_limpio["DistanceFromHome"] = df_limpio["DistanceFromHome"].abs()
df_limpio

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField EmployerNumber  \
0                    6          3               NaN          162,0   
1                    1          4     Life Sciences          259,0   
2                    4          2  Technical Degree          319,0   
3                    2          4           Medical            NaN   
4                    3          3  Technical Degree            NaN   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences          967,0   
1610                 4          2               NaN          972,0   
1611                13          5               NaN            NaN   
1612                 8          4  Technical Degree          990,0   
1613                 7          2           Medical            NaN   

      EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                           1       0         51               3         5   
1                           3       0         65               2         5   
2                           3       0         58               3         5   
3                           1       1         82               3         4   
4                           1       1         45               4         4   
...                       ...     ...        ...             ...       ...   
1609                        3       0         51               2         3   
1610                        3       0         57               3         2   
1611                       13       0         41               4         3   
1612                        1       0         84               3         2   
1613                        4       0         75               3         3   

                       JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0           Research Director                 3           NaN       19537,0   
1                     Manager                 3           NaN       19999,0   
2                     Manager                 4       Married       19232,0   
3           Research Director                 3       Married       17169,0   
4             Sales Executive                 1      Divorced           NaN   
...                        ...              ...           ...           ...   
1609          Sales Executive                 4       Married       10325,0   
1610    Laboratory Technician                 2           NaN        4447,0   
1611          Sales Executive                 3        Single           NaN   
1612          Sales Executive                 4      Divorced           NaN   
1613   Manufacturing Director                 3           NaN       10845,0   

      MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike  \
0            6462                   7       No                 13   
1            5678                   0      NaN                 14   
2            4933                   1       No                 11   
3           26703                   3      NaN                 19   
4            7739                   2       No          

In [26]:
df_limpio["EnvironmentSatisfaction"].value_counts()

EnvironmentSatisfaction
4     460
3     459
1     298
2     297
12      7
35      6
13      6
14      5
24      5
47      5
36      4
48      4
41      4
46      4
42      4
45      3
11      3
22      3
17      3
18      3
20      3
25      3
27      3
37      2
29      2
19      2
38      2
31      2
15      2
16      2
26      1
39      1
10      1
49      1
21      1
28      1
33      1
43      1
Name: count, dtype: int64

In [27]:
df_limpio['EnvironmentSatisfaction'] = df_limpio["EnvironmentSatisfaction"].astype(str)

In [28]:
df_limpio['EnvironmentSatisfaction'].value_counts

<bound method IndexOpsMixin.value_counts of 0        1
1        3
2        3
3        1
4        1
        ..
1609     3
1610     3
1611    13
1612     1
1613     4
Name: EnvironmentSatisfaction, Length: 1614, dtype: object>

In [29]:

df_limpio['EnvironmentSatisfaction'].str.split(expand=True).get([0])

0
0      1
1      3
2      3
3      1
4      1
...   ..
1609   3
1610   3
1611  13
1612   1
1613   4

[1614 rows x 1 columns]

In [31]:
df_limpio["RemoteWork"] = df_limpio["RemoteWork"].str.replace("1", "True").replace("0", "False").replace("Yes", "True")
df_limpio

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField EmployerNumber  \
0                    6          3               NaN          162,0   
1                    1          4     Life Sciences          259,0   
2                    4          2  Technical Degree          319,0   
3                    2          4           Medical            NaN   
4                    3          3  Technical Degree            NaN   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences          967,0   
1610                 4          2               NaN          972,0   
1611                13          5               NaN            NaN   
1612                 8          4  Technical Degree          990,0   
1613                 7          2           Medical            NaN   

     EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                          1       0         51               3         5   
1                          3       0         65               2         5   
2                          3       0         58               3         5   
3                          1       1         82               3         4   
4                          1       1         45               4         4   
...                      ...     ...        ...             ...       ...   
1609                       3       0         51               2         3   
1610                       3       0         57               3         2   
1611                      13       0         41               4         3   
1612                       1       0         84               3         2   
1613                       4       0         75               3         3   

                       JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0           Research Director                 3           NaN       19537,0   
1                     Manager                 3           NaN       19999,0   
2                     Manager                 4       Married       19232,0   
3           Research Director                 3       Married       17169,0   
4             Sales Executive                 1      Divorced           NaN   
...                        ...              ...           ...           ...   
1609          Sales Executive                 4       Married       10325,0   
1610    Laboratory Technician                 2           NaN        4447,0   
1611          Sales Executive                 3        Single           NaN   
1612          Sales Executive                 4      Divorced           NaN   
1613   Manufacturing Director                 3           NaN       10845,0   

      MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike  \
0            6462                   7       No                 13   
1            5678                   0      NaN                 14   
2            4933                   1       No                 11   
3           26703                   3      NaN                 19   
4            7739                   2       No                 12   

In [32]:
df_limpio["NumberChildren"].value_counts()

Series([], Name: count, dtype: int64)

In [33]:
df_limpio = df_limpio.drop(columns="NumberChildren")


In [34]:
df_limpio 

Age Attrition BusinessTravel DailyRate                Department  \
0     51        No            NaN    684,0$                       NaN   
1     52        No            NaN    699,0$                       NaN   
2     42        No  travel_rarely    532,0$   Research & Development    
3     47        No  travel_rarely    359,0$                       NaN   
4     46        No            NaN   1319,0$                       NaN   
...   ..       ...            ...       ...                       ...   
1609  36       Yes  travel_rarely    530,0$                       NaN   
1610  45        No     non-travel    805,0$                       NaN   
1611  39        No  travel_rarely    903,0$                       NaN   
1612  36        No     non-travel   1229,0$                       NaN   
1613  46        No            NaN    566,0$                       NaN   

      DistanceFromHome  Education    EducationField EmployerNumber  \
0                    6          3               NaN          162,0   
1                    1          4     Life Sciences          259,0   
2                    4          2  Technical Degree          319,0   
3                    2          4           Medical            NaN   
4                    3          3  Technical Degree            NaN   
...                ...        ...               ...            ...   
1609                 3          1     Life Sciences          967,0   
1610                 4          2               NaN          972,0   
1611                13          5               NaN            NaN   
1612                 8          4  Technical Degree          990,0   
1613                 7          2           Medical            NaN   

     EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  JobLevel  \
0                          1       0         51               3         5   
1                          3       0         65               2         5   
2                          3       0         58               3         5   
3                          1       1         82               3         4   
4                          1       1         45               4         4   
...                      ...     ...        ...             ...       ...   
1609                       3       0         51               2         3   
1610                       3       0         57               3         2   
1611                      13       0         41               4         3   
1612                       1       0         84               3         2   
1613                       4       0         75               3         3   

                       JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0           Research Director                 3           NaN       19537,0   
1                     Manager                 3           NaN       19999,0   
2                     Manager                 4       Married       19232,0   
3           Research Director                 3       Married       17169,0   
4             Sales Executive                 1      Divorced           NaN   
...                        ...              ...           ...           ...   
1609          Sales Executive                 4       Married       10325,0   
1610    Laboratory Technician                 2           NaN        4447,0   
1611          Sales Executive                 3        Single           NaN   
1612          Sales Executive                 4      Divorced           NaN   
1613   Manufacturing Director                 3           NaN       10845,0   

      MonthlyRate  NumCompaniesWorked OverTime  PercentSalaryHike  \
0            6462                   7       No                 13   
1            5678                   0      NaN                 14   
2            4933                   1       No                 11   
3           26703                   3      NaN                 19   
4            7739                   2       No                 12   

In [35]:
df_limpio["Gender"].replace(to_replace= [0, 1], value=["M", "F"], inplace=True)

In [36]:
df_limpio.head(2)

Age Attrition BusinessTravel DailyRate Department  DistanceFromHome  \
0  51        No            NaN    684,0$        NaN                 6   
1  52        No            NaN    699,0$        NaN                 1   

   Education EducationField EmployerNumber EnvironmentSatisfaction Gender  \
0          3            NaN          162,0                       1      M   
1          4  Life Sciences          259,0                       3      M   

  HourlyRate  JobInvolvement  JobLevel              JobRole  JobSatisfaction  \
0         51               3         5   Research Director                 3   
1         65               2         5             Manager                 3   

  MaritalStatus MonthlyIncome  MonthlyRate  NumCompaniesWorked OverTime  \
0           NaN       19537,0         6462                   7       No   
1           NaN       19999,0         5678                   0      NaN   

   PercentSalaryHike PerformanceRating  RelationshipSatisfaction  \
0                 13               3,0                         3   
1                 14               3,0                         1   

  StandardHours  StockOptionLevel TotalWorkingYeras  TrainingTimesLastYear  \
0           NaN                 0               NaN                      5   
1           NaN                 1              34,0                      5   

  WorkLifeBalance  YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0             3,0              20                NaN                       15   
1             3,0              33                NaN                       11   

   YearsWithCurrManager  DateBirth RoleDepartament RemoteWork  
0                    15       1972             NaN       True  
1                     9       1971             NaN       True

In [37]:
#Variables categóricas
df_limpio.describe(include='object').T

count unique                       top  freq
Age                      1614     54                        35    84
Attrition                1614      2                        No  1355
BusinessTravel            842      3             travel_rarely   586
DailyRate                1614    849                      nan$   124
Department                302      3   Research & Development    196
EducationField            869      6             Life Sciences   349
EmployerNumber           1183   1079                     482,0     2
EnvironmentSatisfaction  1614     38                         4   460
Gender                   1614      2                         M   971
HourlyRate               1614     72             Not Available    84
JobRole                  1614      9          Sales Executive    369
MaritalStatus             963      3                   Married   439
MonthlyIncome             771    668                    6347,0     4
OverTime                  938      2                        No   682
PerformanceRating        1419      2                       3,0  1205
StandardHours             419      1                      80,0   419
TotalWorkingYeras        1088     40                      10,0   144
WorkLifeBalance          1506      4                       3,0   913
YearsInCurrentRole         34     10                       2,0    11
RoleDepartament           302    301        MaNAgeR  -  Sales      2
RemoteWork               1614      2                      True  1000

In [38]:
#Variables numéricas
df_limpio.describe(include=[np.number]).T

count          mean          std     min     25%  \
DistanceFromHome          1614.0     11.141884    10.450560     1.0     2.0   
Education                 1614.0      2.925031     1.022357     1.0     2.0   
JobInvolvement            1614.0      2.739777     0.711567     1.0     2.0   
JobLevel                  1614.0      2.068154     1.101344     1.0     1.0   
JobSatisfaction           1614.0      2.738538     1.106163     1.0     2.0   
MonthlyRate               1614.0  14284.495663  7110.414585  2094.0  8001.0   
NumCompaniesWorked        1614.0      2.673482     2.506152     0.0     1.0   
PercentSalaryHike         1614.0     15.165428     3.648610    11.0    12.0   
RelationshipSatisfaction  1614.0      2.704461     1.079031     1.0     2.0   
StockOptionLevel          1614.0      0.791202     0.842396     0.0     0.0   
TrainingTimesLastYear     1614.0      2.809789     1.297765     0.0     2.0   
YearsAtCompany            1614.0      7.132590     6.124237     0.0     3.0   
YearsSinceLastPromotion   1614.0      2.245973     3.235665     0.0     0.0   
YearsWithCurrManager      1614.0      4.220570     3.562695     0.0     2.0   
DateBirth                 1614.0   1986.076208     9.101332  1963.0  1980.0   

                              50%      75%      max  
DistanceFromHome              8.0     17.0     49.0  
Education                     3.0      4.0      5.0  
JobInvolvement                3.0      3.0      4.0  
JobLevel                      2.0      3.0      5.0  
JobSatisfaction               3.0      4.0      4.0  
MonthlyRate               14248.5  20364.0  26999.0  
NumCompaniesWorked            2.0      4.0      9.0  
PercentSalaryHike            14.0     18.0     25.0  
RelationshipSatisfaction      3.0      4.0      4.0  
StockOptionLevel              1.0      1.0      3.0  
TrainingTimesLastYear         3.0      3.0      6.0  
YearsAtCompany                5.0      9.0     40.0  
YearsSinceLastPromotion       1.0      3.0     15.0  
YearsWithCurrManager          3.0      7.0     17.0  
DateBirth                  1987.0   1993.0   2005.0

In [39]:
## Convertir las columnas a los tipos adecuados. AGE
df_limpio['Age'] = pd.to_numeric(df_limpio['Age'], errors='coerce')# age a numérico. los valores no numéricos se convierten a NaN
df_limpio['Age'] = df_limpio['Age'].astype(int)# age de float a int
df_limpio

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

### Tareas realizadas:
Eliminar columna:
    - Salary - Todos los valores tienen el mismo dato (1000000000$) - Ok en df_limpio
    - SameAsMonthlyIncome - Hemos comprobado que los datos son los mismos que la columna MonthlyIncome - Ok en df_limpio
    - employeecount - Todos los valores son 1 - Ok en df_limpio
    - NUMBERCHILDREN - Todos los valores son nulos. Ok en df_limpio

### Tareas pendientes:
    
1. Cambiar datos:

    1.1. Distancia al trabajo - Cambiar los negativos por positivos (confirmado por negocio)
    
    1.2. RemoteWork - (Unificar datos 0 y false) - (Yes, True y 1)

    1.3. RoleDepartament - Renombrar.
    
    1.4. Age - Pasarlo a integer.

2. Dudas:

    2.1. StandardHours - Todos los valores no nulos son valor unico 80 , 419 registros.Confirmar si los valores nulos corresponden  a jornada completa y los otros a media jornada

    2.2. BusinessTravel - Como interpretamos los nulos?
    
    

